File: generate.php
<?php

$wikis = array(
	'global',
	'cawiki',
	'commonswiki',
	'czwiki',
	'dewiki',
	'enwiki',
	'enwikivoyage',
	'eswiki',
	'etwiki',
	'fiwiki',
	'frwiki',
	'hewiki',
	'huwiki',
	'itwiki',
	'jawiki',
	'kowiki',
	'mediawikiwiki',
	'nlwiki',
	'plwiki',
	'ptwiki',
	'rowiki',
	'ruwiki',
	'skwiki',
	'svwiki',
	'tewiki',
	'thwiki',
	'viwiki',
);

$default_fields = 'event_type, event_total, timestamp, wiki';

$perf_metrics = array(
	array( 'name' => 'userinfo', 'where' => 'event_type = \'userinfo\'', 'fields' => $default_fields ),
	array( 'name' => 'imageinfo', 'where' => 'event_type = \'imageinfo\'', 'fields' => $default_fields ),
	array( 'name' => 'thumbnailinfo', 'where' => 'event_type = \'thumbnailinfo\'', 'fields' => $default_fields ),
	array( 'name' => 'filerepoinfo', 'where' => 'event_type = \'filerepoinfo\'', 'fields' => $default_fields ),
	array( 'name' => 'imageusage', 'where' => 'event_type = \'imageusage\'', 'fields' => $default_fields ),
	array( 'name' => 'globalusage', 'where' => 'event_type = \'globalusage\'', 'fields' => $default_fields ),
	array( 'name' => 'image', 'where' => 'event_type = \'image\'', 'fields' => $default_fields ),
	array( 'name' => 'imagemiss', 'where' => 'event_type = \'image\' AND LENGTH(event_XCache) > 0 AND event_varnish1hits = 0 AND event_varnish2hits = 0 AND event_varnish2hits = 0', 'fields' => 'event_type, event_total, timestamp, wiki, event_XCache, event_varnish1hits, event_varnish2hits, event_varnish3hits' ),
	array( 'name' => 'imagehit', 'where' => 'event_type = \'image\' AND LENGTH(event_XCache) > 0 AND (event_varnish1hits > 0 OR event_varnish2hits > 0 OR event_varnish2hits > 0)', 'fields' => 'event_type, event_total, timestamp, wiki, event_XCache, event_varnish1hits, event_varnish2hits, event_varnish3hits' ),
);

$geo_perf_metrics = array (
	array( 'name' => 'api', 'where' => 'event_type != \'image\'' ),
	array( 'name' => 'image', 'where' => 'event_type = \'image\'' )
);

function generate( $folder, $wiki, $metric = null ) {
	$replacement = '';

	if ( $wiki != 'global' ) {
		$replacement = 'wiki = \'' . $wiki . '\' AND';
	} else {
		$replacement = 'wiki IS NOT NULL AND';
	}

	$sql = file_get_contents( $folder . '/template.sql' );
	$sql = str_replace( '%wiki%', $replacement, $sql );

	if ( !is_null( $metric ) ) {
		$sql = str_replace( '%metricname%', $metric['name'], $sql );
		$sql = str_replace( '%metricfields%', $metric['fields'], $sql );
		$sql = str_replace( '%metricwhere%', $metric['where'], $sql );
	}

	file_put_contents( $folder . '/' . $wiki . ( is_null( $metric ) ? '' : '-'  . $metric['name'] ) . '.sql', $sql );
}

foreach ( $wikis as $wiki ) {
	foreach ( $perf_metrics as $perf_metric) {
		generate( 'perf', $wiki, $perf_metric );
	}

	foreach ( $geo_perf_metrics as $perf_metric) {
		generate( 'geoperf', $wiki, $perf_metric );
	}

	generate( 'actions', $wiki );
}

?>

File: generate.py
#!/usr/bin/env python

import os
import sys

wikis = [
    'global',
    'cawiki',
    'commonswiki',
    'czwiki',
    'dewiki',
    'enwiki',
    'enwikivoyage',
    'eswiki',
    'etwiki',
    'fiwiki',
    'frwiki',
    'hewiki',
    'huwiki',
    'itwiki',
    'jawiki',
    'kowiki',
    'mediawikiwiki',
    'nlwiki',
    'plwiki',
    'ptwiki',
    'rowiki',
    'ruwiki',
    'skwiki',
    'svwiki',
    'tewiki',
    'thwiki',
    'viwiki',
]

default_fields = 'event_type, event_total, timestamp, wiki'

perf_metrics = [
    {'name': 'userinfo', 'where': 'event_type = \'userinfo\'', 'fields': default_fields},
    {'name': 'imageinfo', 'where': 'event_type = \'imageinfo\'', 'fields': default_fields},
    {'name': 'thumbnailinfo', 'where': 'event_type = \'thumbnailinfo\'', 'fields': default_fields},
    {'name': 'filerepoinfo', 'where': 'event_type = \'filerepoinfo\'', 'fields': default_fields},
    {'name': 'imageusage', 'where': 'event_type = \'imageusage\'', 'fields': default_fields},
    {'name': 'globalusage', 'where': 'event_type = \'globalusage\'', 'fields': default_fields},
    {'name': 'image', 'where': 'event_type = \'image\'', 'fields': default_fields},
    {'name': 'imagemiss', 'where': 'event_type = \'image\' AND LENGTH(event_XCache) > 0 AND event_varnish1hits = 0 AND event_varnish2hits = 0 AND event_varnish2hits = 0', 'fields': 'event_type, event_total, timestamp, wiki, event_XCache, event_varnish1hits, event_varnish2hits, event_varnish3hits'},
    {'name': 'imagehit', 'where': 'event_type = \'image\' AND LENGTH(event_XCache) > 0 AND (event_varnish1hits > 0 OR event_varnish2hits > 0 OR event_varnish2hits > 0)', 'fields': 'event_type, event_total, timestamp, wiki, event_XCache, event_varnish1hits, event_varnish2hits, event_varnish3hits'},
]

geo_perf_metrics = [
    {'name': 'api', 'where': 'event_type != \'image\'', 'fields': ''},
    {'name': 'image', 'where': 'event_type = \'image\'', 'fields': ''},
]


def generate(folder, wiki, metric=None):
    replacement = ''

    if wiki != 'global':
        replacement = 'wiki = \'' + wiki + '\' AND'

    sql = open(
        os.path.dirname(os.path.realpath(__file__)) +
        '/' + folder + '/template.sql').read(10000)
    sql = sql.replace('%wiki%', replacement)

    if metric is not None:
        sql = sql.replace('%metricname%', metric['name'])
        sql = sql.replace('%metricfields%', metric['fields'])
        sql = sql.replace('%metricwhere%', metric['where'])

    f = open(
        sys.argv[1] + '/' + folder + '/' +
        wiki + ('' if metric is None else '-' + metric['name']) + '.sql', 'w')
    f.write(sql)
    f.close()

if len(sys.argv) != 2:
    print 'Must specific a folder to send generated SQL to'
    sys.exit(2)

for wiki in wikis:
    for perf_metric in perf_metrics:
        generate('perf', wiki, perf_metric)

    for geo_perf_metric in geo_perf_metrics:
        generate('geoperf', wiki, geo_perf_metric)

    generate('actions', wiki)
